doc/rofi-dmenu.5

```diff
@@ -147,6 +147,13 @@ Or any combination: '5,-3:,7:11,2,0,-9'
 .PP
 Urgent row, mark \fIX\fP as urgent. See \fB\fC-a\fR option for details.
 
+.PP
+\fB\fC-select-rows\fR \fIX\fP
+
+.PP
+If multi-select is enabled, pre-select rows, See \fB\fC-a\fR option for format details.
+If same row is specified multiple times, it state is toggled on subsequential sets.
```

Comment:
I think this behavior is surprising. I prefer the box is checked if the row id appears more than once, and unchecked if the row id doesn't appear.

Is there any reason you think toggling based on odd/even number of occurence a superior behavior?

Reply:
It would make it so much easier to select all rows except  a number of rows...
I am trying to make things easier for the user.

Reply:
I see.

I would have suggested adding an option named `-unselect-rows`, but then I realized it cannot substitute `-select-rows $FOO:$BAR:$BAZ`.

So now the question is: This python-like range syntax is also used in other options as well, do these options also support this toggling behavior? If they do, it is consistent. If they don't, it means that the range syntax is lacking, and instead of adding toggling behavior, we should improve the range syntax (in other PR/discussion ofc).

Reply:
> This python-like range syntax is also used in other options as well, do these options also support this toggling behavior?

There is another option called `-a` to make "active rows". And it **does not have** the toggling behavior. Which means that `-a 1,2,3,2` would make 1, 2, 3 active whilst `-select-rows 1,2,3,2` only checks 1 and 3.

Reply:
Neither '-a' or '-u' document that it has the toggling behaviour. I have no idea why you think it does?
I expect that developers of script that use advanced features like this will read the documentation and can understand the difference.

Reply:
If you say so. I only use rofi for a few use cases, so this quirk won't matter to me that much right now.

If I have sowed doubt in you, then you should ask other developers and users to see what they think. If you are confident that this is best approach, just merge it.

Reply:
???

Reply:
> If anybody is wiling to update the `-u`,`-a` to support toggling that would be appreciated.

Ah. I see what you mean. Although, even if it can be achieved, suddenly add toggling to an existing option that didn't support it before could be a breaking change.

It is true that if the documentation doesn't state that it supports a feature, then it won't. And users should read documentation. However, documentation doesn't make inconsistency disappears from an interface. Take JavaScript for example, `[] + {}` being `"[object Object]"` and `{} + []` being 0 is well documented, but still considered unexpected and inconsistent.

I appreciate that you think ahead and add more features to make the tool more versatile. However, (mis)features are easier added than removed. This is a new feature, it is an opportunity avoid footguns.

How about this? Introduce a new syntax that is superset of the range syntax: `-select-rows range1 & !range2` should select `range1` excluding `range2`. Alternatively, `-select-rows range1 \ range2`. Toggling behavior is of course not needed.

Reply:
> Ah. I see what you mean. Although, even if it can be achieved, suddenly add toggling to an existing option that didn't support it before could be a breaking change.

This option did not exist before, `-select-rows` was added in this PR.
No existing option has been changed, no backwards compatibility with existing options has been broken in this PR.

>How about this? Introduce a new syntax that is superset of the range syntax: -select-rows range1 & !range2 should select range1 excluding range2. Alternatively, -select-rows range1 \ range2. Toggling behavior is of course not needed.

I don't see how this is more consistent. But more important, this  will introduce a shell escaping nightmare as &  and ! have special meaning .

Reply:
> No existing option has been changed, no backwards compatibility with existing options has been broken in this PR.

I was talking about "If anybody is wiling to update the -u,-a to support toggling that would be appreciated."

> this will introduce a shell escaping nightmare as & and ! have special meaning

There is always other syntax beside `&`, `!`, and `\`. The range syntax doesn't support letters, right? so the range expression (let's call it that) can always add a few keywords. For example, `-select-rows 'range1 excluding (range2 excluding range3)` or `-select-rows 'range1 and not(range2 and not range3)'`.

> I don't see how this is more consistent

Because it does not make the range syntax of `-select-rows` to have special meaning of toggling behavior despite other similar options not supporting it. Also, if someone wishes to introduce the range expression syntax to `-a` and `-u` in the future, it would not be a breaking change.

Reply:
I was saying about the consistency because the proposed solution  ('&','!' ) or the new one now ( 'and' and 'not') is not the type of syntax we in rofi.
If this is so important we should not introduce another one.

To specify multiple items, we use ',' separator in several places. 

```css 
entry, prompt {
}
window {
children: [inputbar, listview];
}
```
or 
```bash
rofi -show drun -modes 'drun,window'
```

So if we list multiple ranges we should stick to ',' as list separator.

To negate we use '-' in the search entry box. But this is not ideal for this usecase.
(you can search for 'browser -chrome'  to search all non chrome browser.)

Reply:
> is not the type of syntax we in rofi.

You need to complete this sentence or I wouldn't understand.

> If this is so important we should not introduce another one.
> 
> To specify multiple items, we use ',' separator in several places.

Reusing `,` wouldn't make sense because `,` is for adding items, not for intersecting set of items. In other words, `,` is an `or` operator, different from and `and` operator. `0:4,2:5` is `0,1,2,3,4` whereas `0:4 and 2:5` would be `2,3`.

> If this is so important

It's not important for my use cases right now, neither is the toggling feature, I only brought this up because I dislike the toggling behavior.

Reply:
> To negate we use '-' in the search entry box. But this is not ideal for this usecase.
> (you can search for 'browser -chrome' to search all non chrome browser.)

Because negative numbers are a thing.

This gives me an idea: Many websites have a search syntax similar to GitHub issue (`k1:v1 k2:v2 pattern`). So rofi could potentially introduce a similar syntax: `0:100 not:(10:90 not:40:50)`; or, if you want the `not` keyword to be more easily discernible from normal text: `0:100 @not:(10:90 @not:40:50)`.

Reply:
>>    is not the type of syntax we in rofi.

> You need to complete this sentence or I wouldn't understand.

  is not the type of syntax we use in rofi.

     

> Reusing , wouldn't make sense because , is for adding items, not for intersecting set of items. In other words, , is an or operator, different from and and operator. 0:4,2:5 is 0,1,2,3,4 whereas 0:4 and 2:5 would be 2,3.

eeuh what?? I was talking about adding items not intersecting,  so ',' makes sense for having a list of ranges that should be selected. (As it is now). We use ',' to separate in a list already. How we specify what we then exclude then is not clear.

If we start describing thing in set theory..  then we should use the right syntax for that.. I would be fine with using  ∪,∩, \  (union, intersection, set diff if i remember correctly)  But I think we will have lost most of the developers/users by now.


Anyway I give up. If anybody wants to pick this up and create a PR , I'll reconsider.
I spend more time I have  this already

Reply:
OK. I have created https://github.com/davatorium/rofi/pull/1808 which removes the toggling behavior.

I only proposed a variety of different syntaxes to replace toggling because I dislike it.